ledger/object/record_db.go

```diff
@@ -229,29 +258,92 @@ func setPosition(txn *badger.Txn, recID insolar.ID, position uint32) error {
 	return txn.Set(fullKey, recID.Bytes())
 }
 
-// TruncateHead remove all records after lastPulse
-func (r *RecordDB) TruncateHead(ctx context.Context, from insolar.PulseNumber) error {
-	it := store.NewReadIterator(r.db.Backend(), recordKey(*insolar.NewID(from, nil)), false)
+func (r *RecordDB) truncateRecordsHead(ctx context.Context, from printableKey, km keyMaker) error {
+	it := store.NewReadIterator(r.db.Backend(), from, false)
+	defer it.Close()
+
+	var hasKeys bool
+	for it.Next() {
+		hasKeys = true
+		key := km(it.Key())
+
+		err := r.db.Delete(key)
+		if err != nil {
+			return errors.Wrapf(err, "can't delete key: %s", key.String())
+		}
+
+		inslogger.FromContext(ctx).Debugf("Erased key: %s", key.String())
+	}
+
+	if !hasKeys {
+		inslogger.FromContext(ctx).Infof("No records. Nothing done. Start key: %s", from.String())
+	}
+
+	return nil
+}
+
+func (r *RecordDB) truncatePositionRecordHead(ctx context.Context, from printableKey, km keyMaker, prefix byte) error {
+	it := store.NewReadIterator(r.db.Backend(), from, false)
 	defer it.Close()
 
 	var hasKeys bool
 	for it.Next() {
 		hasKeys = true
-		key := newRecordKey(it.Key())
-		keyID := insolar.ID(key)
+		if it.Key()[0] != prefix {
+			continue
+		}
+		key := km(it.Key())
 
-		err := r.db.Backend().Update(func(txn *badger.Txn) error {
-			fullKey := append(key.Scope().Bytes(), key.ID()...)
-			return txn.Delete(fullKey)
-		})
+		err := r.db.Delete(key)
 		if err != nil {
-			return errors.Wrapf(err, "can't delete key: %+v", key)
+			return errors.Wrapf(err, "can't delete key: %s", key.String())
 		}
-		inslogger.FromContext(ctx).Debugf("Erased key with pulse number: %s. ID: %s", keyID.Pulse().String(), keyID.String())
+
+		inslogger.FromContext(ctx).Debugf("Erased key: %s", key.String())
 	}
 
 	if !hasKeys {
-		inslogger.FromContext(ctx).Infof("No records. Nothing done. Pulse number: %s", from.String())
+		inslogger.FromContext(ctx).Infof("No records. Nothing done. Start key: %s", from.String())
+	}
+
+	return nil
+}
+
+type printableKey interface {
```

Comment:
Why do we need printable key? Use explicit functions instead.

Reply:
I need to recuse function truncatePositionRecordHead